retry on serialization errors

vito opened the pull request:
Fixes https://github.com/dexidp/dex/issues/1341

Turns out it's easy enough to just capture the specific serialization error code and retry the transaction.

Tested this locally - the storage conformance tests pass (they failed when removing serialization altogether), and our integration tests no longer fail on concurrent logins.

srenatus commented on storage/sql/sql.go:
```diff
@@ -59,19 +59,34 @@ var (
-				return err
-			}
-			defer tx.Rollback()
+			for {
```
❓ This is addressing an edge case, but can we limit the number of retries? Also, do we care that it's hammering or does this need some wait times?

vito replied:
Hmmm, to be honest I'm not sure. The Postgres docs just say "retry" but don't say anything about backoff and attempts.

I'm a little nervous about limiting it as it could potentially mean bubbling up the error, given enough concurrency. I agree that it's scary to have it unbounded, but so long as we're getting serialization failures, I'm not sure there's any resolution other than to retry until it works (which should eventually happen as requests calm down). It helps that we're at least catching this very specific error code and only retrying on that one.

Not sure about the hammering/wait times. In my experience it's sufficient to just try again immediately to resolve concurrency "disputes", since they tend to resolve fairly quickly, and the net effect is usually just a few more queries (which the database should be able to handle). If it were a failure that could come from load or network connectivity, I'd totally agree, but this almost feels more like regular handling of database behavior. 🤷‍♂️

srenatus commented on storage/sql/sql.go:
```diff
@@ -59,19 +59,34 @@ var (
-			}
-			defer tx.Rollback()
+			for {
+				tx, err := db.Begin()
```
I wonder if it would be worth switching to `BeginTx(context.TODO(), sql.LevelSerializable)`, which would, I think, mean we can ditch the explicit `SET TRANSACTION ISOLATION LEVEL SERIALIZABLE`.

srenatus replied:
I suppose if we did that we could also go all-in on contexts and do something like
```go
ctx, cancel := context.WithCancel(context.TODO())
defer cancel() // will rollback transaction
tx, err := db.BeginTx(ctx, nil /* or some proper level? */)
if err != nil {
  return err
}
if err := fn(tx); err != nil {
  return err
}
err = tx.Commit()
// etc
```

vito replied:
Neat, didn't know that was a thing! I'll try it out.

vito commented on storage/sql/sql.go:
```diff
@@ -59,19 +59,34 @@ var (
+
+				err = tx.Commit()
+				if err != nil {
+					if pqErr, ok := err.(*pq.Error); ok && pqErr.Code == "40001" {
```
Ah yeah that's much better.

vito commented:
@srenatus Yeah, I'm a bit confused by it too. Maybe the failures have something to do with concurrent reads, not just writes? For example if there's any other query that runs against that table at the same time someone may be logging in - maybe auth request garbage-collection?

vito commented on storage/sql/sql.go:
```diff
@@ -59,19 +60,35 @@ var (
+				}
+
+				if err := fn(tx); err != nil {
+					return err
```
Ack, I think this error also needs to be checked. Our integration tests just failed, so maybe it was luck that it passed for me. I think the `fn` here returns an error because whatever query it's running sees the serialization failure. Unfortunately the `fn` also sometimes wraps the error, making it impossible to check the error code.

vito commented:
@srenatus Alright, I've pushed another commit that refactors all the error wrapping out of the `ExecTx` fn, and I left a comment warning future developers to not wrap errors in there. Without this last commit, my initial fix only applied some of the time, since the failure could either happen on the individual queries in `fn` or the `Commit` at the end, which was the only thing I added the check for. =

Along the way I tried to keep the error wrapping/handling consistent across all the `conn` methods. This change ended up being kinda big, but if it makes you feel any better there *were* tests that failed along the way, so the test coverage seems decent. I didn't end up having to change any tests, since no user-facing behavior changed once I fixed the things the tests told me to.

With this all cleaned up, I've got our integration tests successfully running in a loop with ~23 concurrent logins.

ericchiang commented on storage/sql/sql.go:
```diff
@@ -51,27 +52,49 @@ var (
+
+				if err := fn(tx); err != nil {
+					if pqErr, ok := err.(*pq.Error); ok && pqErr.Code.Name() == "serialization_failure" {
+						// serialization error; retry
```
In the initial bug report you mentioned

```
The Postgres docs say "applications using this level must be prepared to retry transactions due to serialization failures", which Dex doesn't seem to be doing, and it doesn't feel right to force the client to retry API/login requests due to this, and I don't think they'd even have enough information to recognize that it's a serialization failure to retry in the first place.
```

Can you add a link to those docs here?

vito replied:
Sure!

srenatus commented on glide.yaml:
```diff
@@ -114,7 +114,7 @@ import:
   version: 3fb7a0e792edd47bf0cf1e919dfc14e2be412e15
 - package: github.com/lib/pq
-  version: 50761b0867bd1d9d069276790bcd4a3bccf2324a
+  version: 9eb73efc1fcc404148b56765b0d3f61d9a5ef8ee
```
❓ There's a lot going on between [these two commits](https://github.com/lib/pq/compare/50761b0867bd1d9d069276790bcd4a3bccf2324a...9eb73efc1fcc404148b56765b0d3f61d9a5ef8ee) -- is there anything in particular that you're after, or is this a chore-type update?

vito replied:
This was necessary for `BeginTx` with transaction level support: https://github.com/lib/pq/commit/ed1a26d9e1fd4eb1cc491e6d688467f4e81a413c

srenatus commented on storage/sql/sql.go:
```diff
@@ -74,6 +75,11 @@ var (
 				}
 
 				if err := fn(tx); err != nil {
+					if pqErr, ok := err.(*pq.Error); ok && pqErr.Code.Name() == "serialization_failure" {
```
💭 So, the `serialization_failure` error _potentially_ comes back from the `fn(tx)`, and not only from the `Commit` below, as previously thought? Also, do we know if it _ever_ comes back from the `tx.Commit()`?

vito replied:
Yes, I saw it come from both. I believe it just comes down to timing.

srenatus commented:
@vito thanks for hanging on to this 👍 looks like in the meantime, some other PRs have gotten into this PR's way (conflicts).

vito commented:
@srenatus Yep, just working on fixing the dependencies up now and adding a comment per @ericchiang's suggestion. Will have it ready soon.

vito commented:
OK, all rebased and good to go!

srenatus commented:
ℹ️ This is probably good to go, but I'd like to take a dive into PG doc land to wrap my head around this some more. However, I'll do that tomorrow morning, probably. In the mean time, there's not going to be a release out of thin air, so I'm fine either way -- shall we merge this now or later?

vito commented:
@srenatus Doesn't matter to me, feel free to take your time. We've still got a few things on our fork so we won't be blocked on a merge either way.
